### PR TITLE
8284553: Deprecate the DEFAULT static field of OAEPParameterSpec

### PR DESCRIPTION
**Pull request by valeriepeng**
This trivial change is to deprecate the DEFAULT static field of OAEPParameterSpec class. Wordings are mostly the same as the previous PSSParameterSpec deprecation change. Rest are just minor code re-factoring.

The CSR will be filed once review is somewhat finished.

Thanks,
Valerie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8284553](https://bugs.openjdk.java.net/browse/JDK-8284553): Deprecate the DEFAULT static field of OAEPParameterSpec
 * [JDK-8284919](https://bugs.openjdk.java.net/browse/JDK-8284919): Deprecate the DEFAULT static field of OAEPParameterSpec (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.java.net/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to a760fc5148eaf04cd4b82a738c3315831446735d

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8191/head:pull/8191` \
`$ git checkout pull/8191`

Update a local copy of the PR: \
`$ git checkout pull/8191` \
`$ git pull https://git.openjdk.java.net/jdk pull/8191/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8191`

View PR using the GUI difftool: \
`$ git pr show -t 8191`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8191.diff">https://git.openjdk.java.net/jdk/pull/8191.diff</a>

</details>

### CODE REVIEW

**Comment by seanjmullan** [comment]
> I think deprecating DEFAULT? is wrong.? RFC8017 still has this definition:
> 
> > RSAES-OAEP-params ::= SEQUENCE {
> > hashAlgorithm      [0] HashAlgorithm     DEFAULT sha1,
> > maskGenAlgorithm   [1] MaskGenAlgorithm  DEFAULT mgf1SHA1,
> > pSourceAlgorithm   [2] PSourceAlgorithm  DEFAULT pSpecifiedEmpty
> > }
> 
> and DEFAULT is what you should be getting if you see an empty sequence in the param field.? It's DEFAULT in ASN1 terms, not DEFAULT in terms of what you should use going forward? to create signatures, and the ASN1 DEFAULT won't change.
> 
> In any event, I can't find where RFC8017 says anything about deprecating the defaults.? AFAICT, although there's general guidance to go away from SHA1,? the math suggests that SHA1 is still sufficient for OAEP,? and there's no guidance specific to OAEP's use of SHA1 that I can find other than the requirement in NIST SP800-56B rev 2 to use "Approved Hash Functions" for OAEP. If there's a section in 8017 that you're looking at for this guidance that I missed, you may want to update your comment to point to it.

[https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.1) (which defines the OAEP-params structure) says:

> o  hashAlgorithm identifies the hash function.  It SHALL be an
>       algorithm ID with an OID in the set OAEP-PSSDigestAlgorithms.  For
>       a discussion of supported hash functions, see [Appendix B.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1).

[https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1) is a long section discussing hash functions, but I think this text (last two paragraphs) is most relevant:

>   There have also been advances in the cryptanalysis of SHA-1.
>    Particularly, the results of Wang et al.  [[SHA1CRYPT](https://www.rfc-editor.org/rfc/rfc8017#ref-SHA1CRYPT)] (which have
>    been independently verified by M.  Cochran in his analysis [[COCHRAN](https://www.rfc-editor.org/rfc/rfc8017#ref-COCHRAN)])
>    on using a differential path to find collisions in SHA-1, which
>    conclude that the security strength of the SHA-1 hashing algorithm is
>    significantly reduced.  However, this reduction is not significant
>    enough to warrant the removal of SHA-1 from existing applications,
>    but its usage is only recommended for backwards-compatibility
>    reasons.
> 
>    To address these concerns, only SHA-224, SHA-256, SHA-384, SHA-512,
>    SHA-512/224, and SHA-512/256 are RECOMMENDED for new applications.
>    As of today, the best (known) collision attacks against these hash
>    functions are generic attacks with complexity 2L/2, where L is the
>    bit length of the hash output.  For the signature schemes in this
>    document, a collision attack is easily translated into a signature
>    forgery.  Therefore, the value L / 2 should be at least equal to the
>    desired security level in bits of the signature scheme (a security
>    level of B bits means that the best attack has complexity 2B).  The
>    same rule of thumb can be applied to RSAES-OAEP; it is RECOMMENDED
>    that the bit length of the seed (which is equal to the bit length of
>    the hash output) be twice the desired security level in bits.

Also, this is a proposed deprecation ... there is no plan to remove this field. The main intention of this deprecation is to alert users that SHA-1 is being used as it may not be apparent from the "DEFAULT" name, and to make sure that is ok or to think about using something stronger.

@valeriepeng I recommend making the link more specific in the deprecation text to point to https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java [hunk 1]
```diff
@@ -72,13 +72,6 @@
  * }
  * EncodingParameters ::= OCTET STRING(SIZE(0..MAX))
  * </pre>
- * <p>Note: the OAEPParameterSpec.DEFAULT uses the following:
- * <pre>
- *     message digest  -- "SHA-1"
- *     mask generation function (mgf) -- "MGF1"
- *     parameters for mgf -- MGF1ParameterSpec.SHA1
- *     source of encoding input -- PSource.PSpecified.DEFAULT
- * </pre>
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:81]: I think you should leave these lines in, so users can still see what the default parameters are. Nit, change:
"the OAEPParameterSpec.DEFAULT uses the following"
to:
"{@code OAEPParameterSpec.DEFAULT} uses the following default values:"

**Comment by valeriepeng** [comment]
> _Mailing list message from [Michael StJohns](mailto:mstjohns@comcast.net) on [security-dev](mailto:security-dev@mail.openjdk.java.net):_
> 
> On 4/11/2022 9:34 PM, Valerie Peng wrote:
> 
> Hi Valerie -
> 
> I think deprecating DEFAULT? is wrong.? RFC8017 still has this definition:
> 
> > RSAES-OAEP-params ::= SEQUENCE {
> > hashAlgorithm      [0] HashAlgorithm     DEFAULT sha1,
> > maskGenAlgorithm   [1] MaskGenAlgorithm  DEFAULT mgf1SHA1,
> > pSourceAlgorithm   [2] PSourceAlgorithm  DEFAULT pSpecifiedEmpty
> > }
> 
> and DEFAULT is what you should be getting if you see an empty sequence in the param field.? It's DEFAULT in ASN1 terms, not DEFAULT in terms of what you should use going forward? to create signatures, and the ASN1 DEFAULT won't change.
> 
> In any event, I can't find where RFC8017 says anything about deprecating the defaults.? AFAICT, although there's general guidance to go away from SHA1,? the math suggests that SHA1 is still sufficient for OAEP,? and there's no guidance specific to OAEP's use of SHA1 that I can find other than the requirement in NIST SP800-56B rev 2 to use "Approved Hash Functions" for OAEP. If there's a section in 8017 that you're looking at for this guidance that I missed, you may want to update your comment to point to it.
> 
> Take care - Mike
> 
> -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.java.net/pipermail/security-dev/attachments/20220411/aaa5d423/attachment.htm>

Hi Mike,
Thanks for the comment/feedback. Deprecating the static field in javadoc does not mean we are changing its parameters with something other than those in RFC 8017. Providers can and should still follow the default ASN.1 values when the DER encoding omit the particular field(s). Besides what Sean mentioned already, this deprecation is an effort to moving toward the "explicitly specify what you want to use" direction. Hope this clarifies thing up?
I like the suggestion for using a more specific link and will change it to point to the particular section. 

Regards,
Valerie

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java (hunk 1 above)
Comment (valeriepeng) [reply src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:81]: Leaving this in meaning we may have to duplicate the new wordings for the field here. The default values are covered in the paragraph above where the ASN.1 structure for RSAES-OAEP-params is defined. I can move this down to the static field DEFAULT.

Comment (seanjmullan) [reply src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:81]: Ok.

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java [hunk 2]
```diff
@@ -89,39 +82,54 @@
...
+    public static final OAEPParameterSpec DEFAULT = new OAEPParameterSpec(
+            "SHA-1", "MGF1", MGF1ParameterSpec.SHA1,
+            PSource.PSpecified.DEFAULT);
 
-    /**
-     * Constructs a parameter set for OAEP padding as defined in
-     * the PKCS #1 standard using the default values.
-     */
+    // disallowed
     private OAEPParameterSpec() {
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:113]: I think you can just remove this ctor now that it is not used by `DEFAULT`.

**Comment by valeriepeng** [comment]
CSR filed at: https://bugs.openjdk.java.net/browse/JDK-8284919

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java [hunk 3]
```diff
@@ -89,39 +82,54 @@
...
+     * The OAEP parameter set with all default values, i.e. "SHA-1" as message
+     * digest algorithm, "MGF1" as mask generation function (mgf) algorithm,
+     * {@code MGF1ParameterSpec.SHA1} as parameters for the mask generation
+     * function, and {@code PSource.PSpecified.DEFAULT} as the source of the
+     * encoding input.
+     *
+     * @deprecated This field uses the default values defined in the PKCS #1
+     *         standard. Some of these defaults are no longer recommended due
+     *         to advances in cryptanalysis -- see the
+     *         <a href="https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1">Appendix B.1</a>
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:100]: In the URL text, you should still say this is from RFC 8017. Suggest changing these two lines to:
     *         to advances in cryptanalysis -- see
     *         <a href="https://www.rfc-editor.org/rfc/rfc8017#appendix-B.1">Appendix B.1 of RFC 8017</a>

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java [hunk 4]
```diff
@@ -89,39 +82,49 @@
...
-     * the PKCS #1 standard using the default values.
+     * The OAEP parameter set with all default values, i.e. "SHA-1" as message
+     * digest algorithm, "MGF1" as mask generation function (mgf) algorithm,
+     * {@code MGF1ParameterSpec.SHA1} as parameters for the mask generation
+     * function, and {@code PSource.PSpecified.DEFAULT} as the source of the
+     * encoding input.
+     *
+     * @deprecated This field uses the default values defined in the PKCS #1
+     *         standard. Some of these defaults are no longer recommended due
+     *         to advances in cryptanalysis -- see the
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:99]: Remove "the".

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java [hunk 5]
```diff
@@ -89,39 +82,49 @@
...
-     * @param pSrc the source of the encoding input P.
-     * @exception NullPointerException if <code>mdName</code>,
-     * <code>mgfName</code>, or <code>pSrc</code> is null.
+     * @param mdName the algorithm name for the message digest
+     * @param mgfName the algorithm name for the mask generation function
+     * @param mgfSpec the parameters for the mask generation function;
+     * if {@code null} is specified, {@code null} will be returned by
+     * {@link #getMGFParameters()}
+     * @param pSrc the source of the encoding input P
+     * @exception NullPointerException if {@code mdName},
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:126]: Change `@exception` to the more accepted `@throws` tag.

**Review by seanjmullan — APPROVED** [review]

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java (hunk 4 above)
Comment (valeriepeng) [reply src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:99]: Ok~

File: src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java (hunk 5 above)
Comment (valeriepeng) [reply src/java.base/share/classes/javax/crypto/spec/OAEPParameterSpec.java:126]: Yes, will change.

**Pull request closed without merging**